qiskit_experiments/database_service/db_experiment_data.py

```diff
@@ -1692,3 +1706,107 @@ def __getattr__(self, name: str) -> Any:
         except KeyError:
             # pylint: disable=raise-missing-from
             raise AttributeError("Attribute %s is not defined" % name)
+
+    def _safe_serialize_jobs(self):
+        """Return serializable object for stored jobs"""
+        # Since Job objects are not serializable this removes
+        # them from the jobs dict and returns {job_id: None}
+        # that can be used to retrieve jobs from a service after loading
+        jobs = ThreadSafeOrderedDict()
+        for jid, _ in self._jobs.items():
```

Comment:
We don't need to lock the container here? What happens if this is called when the container is locked by other thread? What happens if we call cancel method after this method is called? Likely job is removed from the container and job.cancel crashes the cancel method.

Reply:
This doesn't remove jobs, it is creating a new copy with just the job ids for the saved data because Job objects are not pickleable nor JSON serializable. The actual job objects could be re-retrieved after loading if the correct backend was reattached (which isn't automatic since IBMQ backends aren't roundtrip serializable).

So I don't think its necessary to lock, but I could change the iter to `.keys` instead of `.items` which implicitly creates a copy to list which means if the dict is somehow changed it won't effect it.